### PR TITLE
fix(master coredump thread): Coredump thread should not retry forever on exception

### PR DESCRIPTION
**Pull request by juliayakovlev**
If exceptions will happen 10 times during getting backtraces - fail the thread and don't run it
forever
[Task](https://trello.com/c/mAFO7cKT/2391-master-coredump-thread-should-not-retry-forever-on-unknown-exception)

## PR pre-checks (self review)

- [x] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [x] I didn't leave commented-out/debugging code
- [x] I added the relevant `backport` labels
- [x] ~~New configuration option are added and documented (in `sdcm/sct_config.py`)~~
- [x] ~~I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)~~
- [x] All new and existing unit tests passed (CI)
- [x] I have updated the Readme/doc folder accordingly (if needed)

### CODE REVIEW

**Review by dkropachev — APPROVED** [review]
LGTM

File: sdcm/coredump.py [hunk 1]
```diff
@@ -83,11 +83,13 @@ def update(self,
...
     lookup_period = 30
     upload_retry_limit = 3
+    # If exceptions will happen MAX_COREDUMP_THREAD_EXCEPTIONS times - fail the thread and don't run it forever
 
-    def __init__(self, node: 'BaseNode', max_core_upload_limit: int):
+    def __init__(self, node: 'BaseNode', max_core_upload_limit: int, max_coredump_thread_exceptions: int = 10):
         self.node = node
         self.log = SDCMAdapter(node.log, extra={"prefix": self.__class__.__name__})
         self.max_core_upload_limit = max_core_upload_limit
+        self.max_coredump_thread_exceptions = max_coredump_thread_exceptions
```
Comment (enaydanov) [inline sdcm/coredump.py:92]: Can we make it a class variable?

File: sdcm/coredump.py [hunk 2]
```diff
@@ -103,11 +105,25 @@ def run(self):
...
+                last_exc = exc
+
+            # If exceptions will happen MAX_COREDUMP_THREAD_EXCEPTIONS times - fail the thread and don't run it forever
+            if exceptions_count == self.max_coredump_thread_exceptions:
+                self.main_cycle_body_reach_exception_limit(last_exc)
+
+    # The function is added for unit test ability to recognize that exception was raised
+    @staticmethod
+    def main_cycle_body_reach_exception_limit(exc):
+        raise exc
```
Comment (enaydanov) [inline sdcm/coredump.py:126]: What about having attrubute `self.exception` for storing raised exception? It'll simplify code a little bit and in result we don't need to do something special for unit tests too:

```python
def __init__(self):
    ...
    self.exception = None
    ...

def run(self):
    exceptions_count = 0
    while ...:
        try:
            self.main_cycle_body()
            exceptions_count = 0
        except Exception as exc:
            self.log.error("...: %s", exc)
            exceptions_count += 1
            if exceptions_count == self.max_coredump_thread_exceptions:
                self.exception = exc
                raise
```

And in unit tests data you can replace list `raised_exceptions` with `exception`.

File: sdcm/coredump.py [hunk 3]
```diff
@@ -163,7 +179,6 @@ def _process_coredumps(
             except:  # pylint: disable=bare-except
                 pass
 
-    @retrying(n=10, sleep_time=20, allowed_exceptions=NETWORK_EXCEPTIONS, message="Retrying on getting pid of cores")
     def get_list_of_cores(self) -> Optional[List[CoreDumpInfo]]:
         return self._get_list_of_cores()
```
Comment (enaydanov) [inline sdcm/coredump.py:183]: If we removed `@retrying` then we don't need this wrapper at all? Probably, we need to rename `_get_list_of_cores()` to `get_list_of_cores()`?

File: unit_tests/test_coredump.py [hunk 4]
```diff
@@ -22,16 +22,17 @@ def wait_ssh_up(self, verbose=False):
 class CoredumpExportSystemdTestThread(CoredumpExportSystemdThread):
     lookup_period = 0
 
-    def __init__(self, node: 'BaseNode', max_core_upload_limit: int):
+    def __init__(self, node: 'BaseNode', max_core_upload_limit: int, max_coredump_thread_exceptions: int):
         self.got_cores = []
-        super().__init__(node, max_core_upload_limit)
+        self.raised_exceptions = []
+        super().__init__(node, max_core_upload_limit, max_coredump_thread_exceptions)
```
Comment (enaydanov) [inline unit_tests/test_coredump.py:28]: See comments above about class variable and `raised_exceptions` -> `exception`.

**Review by enaydanov — CHANGES_REQUESTED** [review]

File: sdcm/coredump.py [hunk 5]
```diff
@@ -83,11 +83,13 @@ def update(self,
 class CoredumpThreadBase(Thread):
     lookup_period = 30
     upload_retry_limit = 3
+    # If exceptions will happen MAX_COREDUMP_THREAD_EXCEPTIONS times - fail the thread and don't run it forever
```
[the remark this answers is not in the document]
Comment (juliayakovlev) [reply sdcm/coredump.py:86]: Removed

File: sdcm/coredump.py (hunk 1 above)
Comment (juliayakovlev) [reply sdcm/coredump.py:92]: This variable is changed during unit tests. Class variable is protected and can't be changed

Comment (juliayakovlev) [reply sdcm/coredump.py]: fixed

File: sdcm/coredump.py (hunk 2 above)
Comment (juliayakovlev) [reply sdcm/coredump.py]: changed

File: unit_tests/test_coredump.py (hunk 4 above)
Comment (juliayakovlev) [reply unit_tests/test_coredump.py]: changed

File: sdcm/coredump.py (hunk 3 above)
Comment (juliayakovlev) [reply sdcm/coredump.py]: fixed

**Review by enaydanov — APPROVED** [review]
LGTM

**Pull request merged**